### PR TITLE
fix/profile reconcile

### LINKED ISSUE #637: sfpowerkit:source:profile:reconcile -s fails with error: Cannot read property 'getUsername' of undefined
**Describe the bug**
`sfdx sfpowerkit:source:profile:reconcile -s` failing with error:

```
-------------------------------------------------------------------------------------------
sfpowerkit  -- The DX@Scale Developer Toolkit -Version:4.1.5 -Release:February 22
-------------------------------------------------------------------------------------------
INFO [2022-02-16 11:29:54.769 +0000] (sfpowerkit): Profiles Found in Entire Drirectory 24
INFO [2022-02-16 11:29:54.769 +0000] (sfpowerkit): Initiated Profile reconcile thread :1  with a chunk of 1 profiles
INFO [2022-02-16 11:29:54.769 +0000] (sfpowerkit): Profiles queued in thread :1 :
INFO [2022-02-16 11:29:54.769 +0000] (sfpowerkit): ["C:\\xx\\force-app\\main\\default\\profiles\\Admin.profile-meta.xml"]
ERROR [2022-02-16 11:29:54.770 +0000] (sfpowerkit): Cannot read property 'getUsername' of undefined
    err: {
      "type": "TypeError",
      "message": "Cannot read property 'getUsername' of undefined",
      "stack":
          TypeError: Cannot read property 'getUsername' of undefined
              at C:\Users\x\AppData\Local\sfdx\node_modules\sfpowerkit\src\impl\source\profiles\profileReconcile.ts:100:33
              at new Promise (<anonymous>)
              at ProfileReconcile.runWorkers (C:\Users\x\AppData\Local\sfdx\node_modules\sfpowerkit\src\impl\source\profiles\profileReconcile.ts:73:25)
              at ProfileReconcile.<anonymous> (C:\Users\x\AppData\Local\sfdx\node_modules\sfpowerkit\src\impl\source\profiles\profileReconcile.ts:58:18)
              at Generator.next (<anonymous>)
              at fulfilled (C:\Users\x\AppData\Local\sfdx\node_modules\sfpowerkit\lib\impl\source\profiles\profileReconcile.js:24:58)

```

**Expected behavior**
Version 4.0.3 runs fine without any problems. 4.1.3 to 4.1.5 are having the error.

**Desktop:**
- OS: Windows 10
- Node Version 14.17.0; 14.18.2
- Version of sfpowerkit 4.1.5
- sfdx-cli/7.137.1 win32-x64 node-v14.18.2

### LINKED ISSUE #655: sfpowerkit:source:profile:reconcile --source-only does not work with parallel workers
**Describe the bug**
Source only mode for `sfpowerkit:source:profile:reconcile` should reconcile profiles based on the given source folder. However after switching to parallel workers (#629 ) this no longer works and the command clears all profiles.
This is mentioned in #637 but the PR only resolves the error and does not restore the previous behaviour.

**To Reproduce**

1. checkout a project with a some metadata that is listed in a profile
2. delete the metadata from the source folder
3. run `sfdx sfpowerkit:source:profile:reconcile -f ${sourcefolder} -s`

**Expected behavior**
All profiles should no longer contain the locally deleted metadata

**Actual behavior**
All profiles are fully cleared (expect name `loginIpRanges` and `custom`)

**Desktop:**

-   OS: Windows/gitBash
-   Node Version: 14.16.1
-   Version of sfpowerkit: 4.2.1

**Additional context**
Switching back to sfpowerkit 4.0.3 or commit 9c43f12e61abdb6d13ed6e2490d9d1e1f3664ae2 resolves the issue.

I looked a bit into it and it I think the problem is, that the workers do not fill the `METADATA_INFO` with the components from the source folder. 

Possible solutions I see:
* don't use the workers for source-only mode as there are no call outs and the speed up should be negligible.
* fill the `METADATA_INFO` with the local components for every worker.

I'm not sure which would be the better one, but I can try to look into one of them.

### PR DESCRIPTION
**Pull request by azlam-abdulsalam**
- fix(reconcile): fix profile reconcile crash when used with source
- refactor to use SDR lib for figuring local components
- refactor variables across profile impl to make it more readable
- fixes #637 and fixes #655

### CODE REVIEW

**Comment by vuha-acn** [comment]
@azlam-abdulsalam - I'll have the team test and report back any issues/discrepancies

**Comment by vuha-acn** [comment]
@azlam-abdulsalam  - Some minor issues discovered during testing

- Log Format needs to carry over old features that includes Timestamp, and color-scheme outputs during processing
- Addition of the `sfpowerkit-cache.db` needs to be added to release notes and the .gitignore file on the dxatscale-template
- Fix formatting of header for the `sfpowerkit -- The DX@Scale Developer Toolkit - Version: 4.2.3 - Release: March 22` (Update spacing and potentially update the Release Date to April 22)

File: src/impl/source/profiles/profileActions.ts [hunk 1]
```diff
@@ -1,103 +1,87 @@
...
-                        profileComponent,
-                        METADATA_INFO.Profile.sourceExtension
-                    );
-                    if (profileName === retrievedProfileFileName && profiles.includes(profileName)) {
-                        profilesStatus.updated.push(profileComponent);
+                for (let j = 0; j < localProfiles.length; j++) {
+                    if (profileName === localProfiles[j].name && remoteProfiles.includes(profileName)) {
+                        profilesStatus.updated.push(localProfiles[j]);
                         found = true;
-                        break;
```
Comment (genoud) [inline src/impl/source/profiles/profileActions.ts:61]: Is it ok to remove this line?

**Review by genoud — COMMENTED** [review]
Hi @azlam-abdulsalam 
I tested the command and it looks good.
I added one comment on one file.

Would it be possible to  use the @salesforce.source-deploy-retrieve package to completely remove dependecy on this file resources\metadatainfo.json?

Thanks

**Comment by azlam-abdulsalam** [comment]
> @azlam-abdulsalam - Some minor issues discovered during testing
> 
> * Log Format needs to carry over old features that includes Timestamp, and color-scheme outputs during processing
> * Addition of the `sfpowerkit-cache.db` needs to be added to release notes and the .gitignore file on the dxatscale-template
> * Fix formatting of header for the `sfpowerkit -- The DX@Scale Developer Toolkit - Version: 4.2.3 - Release: March 22` (Update spacing and potentially update the Release Date to April 22)

- Log format was changed as part of the last release
- Formatting of header will be fixed in a seperate PR, as this one doesnt touch the component
- I have moved the location of sfpowerkit-cache.db to a temporary Directory

File: src/impl/source/profiles/profileActions.ts (hunk 1 above)
Comment (azlam-abdulsalam) [reply src/impl/source/profiles/profileActions.ts:61]: yeah, this is for a use case where you have the same profile (reconciled ) ones across different packages.

**Comment by azlam-abdulsalam** [comment]
> Hi @azlam-abdulsalam I tested the command and it looks good. I added one comment on one file.
> 
> Would it be possible to use the @salesforce.source-deploy-retrieve package to completely remove dependecy on this file resources\metadatainfo.json?
> 
> Thanks

Thanks @genoud for the tests

File: src/impl/source/profiles/profileActions.ts (hunk 1 above)
Comment (azlam-abdulsalam) [reply src/impl/source/profiles/profileActions.ts:61]: > Hi @azlam-abdulsalam I tested the command and it looks good. I added one comment on one file.
> 
> Would it be possible to use the @salesforce.source-deploy-retrieve package to completely remove dependecy on this file resources\metadatainfo.json?
> 
> Thanks

I think its possible to completely remove medatainfo.json, but there are other areas such as diff, that need to be reworked

**Pull request merged**